diffkemp/simpll/DifferentialFunctionComparator.cpp

```diff
@@ -445,28 +452,26 @@ int DifferentialFunctionComparator::cmpAllocs(const CallInst *CL,
         return 0;
     }
 
-    // The instruction is a call instrution calling the function kzalloc. Now
-    // look whether the next instruction is a BitCastInst casting to a structure
-    // type.
-    if (!isa<BitCastInst>(CL->getNextNode())
-        || !isa<BitCastInst>(CR->getNextNode()))
-        return 1;
-
     // Check if kzalloc has constant size of the allocated memory
     if (!isa<ConstantInt>(CL->getOperand(0))
         || !isa<ConstantInt>(CR->getOperand(0)))
         return 1;
 
-    // Get the allocated structure types
-    StructType *STyL = getStructType(CL->getNextNode());
-    StructType *STyR = getStructType(CR->getNextNode());
-
-    // Return 0 (equality) if both allocated types are structs of the same name
-    // and each struct has a size equal to the size of the allocated memory.
-    return !STyL || !STyR
-           || cmpStructTypeSizeWithConstant(STyL, CL->getOperand(0))
-           || cmpStructTypeSizeWithConstant(STyR, CR->getOperand(0))
-           || STyL->getStructName() != STyR->getStructName();
+    // If the next instruction is a bitcast, compare its type instead
+    const Value *ValL =
+            isa<BitCastInst>(CL->getNextNode()) ? CL->getNextNode() : CL;
+    const Value *ValR =
+            isa<BitCastInst>(CR->getNextNode()) ? CR->getNextNode() : CR;
```

Comment:
Could this piece of code be moved into `getPointeeStructTypeInfo`? IIUC, it's a part of the type info extraction - if the following instruction is a bitcast, then the type info is extracted from it.

Reply:
It is a part of type info extraction, but specific to `cmpAllocs`. `cmpMemset` also uses `getPointeeStructTypeInfo`, but it works with bitcasts in the "opposite way" - if the instruction is a bitcast, it extracts the type from the source value.

I would leave it here just to avoid errors in a scenario where `cmpMemset` calls `getPointeeStructTypeInfo` and there is some unrelated bitcast that follows the instruction.

Reply:
Right, good point.